### PR TITLE
Objectron jpeg

### PR DESCRIPTION
**Pull request by rjean**
- 16-bit precision training on GPU (33% faster training on 2080 Ti - Lower memory usage on all hardware -> Bigger batch size.
- Online Classification Accuracy Evaluation using Nearest Neighbors
- JPEG FIle-base dataloader
- Different learning rate for predictor and the rest.
- Different formulation of the loss function.
- Script to generate embeddings used for evaluation
- Change predictor / projector size and number of layers.
- Dry-run mode for quick regression testing

### CODE REVIEW

File: selfsupmotion/models/simsiam.py [hunk 1]
```diff
@@ -68,54 +192,101 @@ def __init__(
...
+        backbone_network = backbone(first_conv=self.first_conv, maxpool1=self.maxpool1, return_all_feature_maps=False)
         self.online_network = SiameseArm(
-            encoder, input_dim=self.hidden_mlp, hidden_size=self.hidden_mlp, output_dim=self.feat_dim
+            backbone_network, input_dim=self.hidden_mlp, hidden_size=self.hidden_mlp, output_dim=self.feat_dim
         )
+        #max_batch = math.ceil(self.num_samples/self.batch_size)
+        encoder, projector = self.online_network.encoder, self.online_network.projector
+        self.feature_dim = self.online_network.encoder.fc.in_features
+        self.train_features = torch.zeros((self.num_samples, self.feature_dim))
+        self.train_meta = []
```
Comment (plstcharles) [inline selfsupmotion/models/simsiam.py:215]: garder tous les embeddings en mémoire, c'est pas dangereux que "boom la RAM"?

File: selfsupmotion/models/simsiam.py [hunk 2]
```diff
@@ -68,54 +192,101 @@ def __init__(
...
+            self.cuda_train_features = self.train_features.cuda()
+
         loss = self.cosine_similarity(h1, z2) / 2 + self.cosine_similarity(h2, z1) / 2
 
+        self.valid_meta+=uid
+        base = batch_idx*self.batch_size
+
+        valid_features = F.normalize(f1, dim=1).detach()
+
+        similarity = torch.mm(valid_features, self.cuda_train_features.T)
```
Comment (plstcharles) [inline selfsupmotion/models/simsiam.py:279]: qqchose me dit que cette opération va devenir coûteuse aussi, on devrait probablement la mettre dans un "if" configurable on/off

**Comment by plstcharles** [comment]
Not bad, not bad! les changements sont pas trop invasifs, du moins je pense tout comprendre!

Question du jour: à quel point veux-tu que je "peer review" le contenu? ya des problèmes mineurs à 2-3 endroits, mais rien qui m'empêcherait de travailler (je contournerais ou je ferais des fixups rapides). C'est vraiment une question de "à quel point on veut que le code soit propre vs à quel point on est pressé d'arriver à des résultats". Dans tous les cas, je vais le rouler réellement demain pendant la journée pour tester le comportement, et on pourra penser à merger si tout est beau.

Aussi: est-ce qu'il y avait des grosses différences d'implémentation entre les deux architectures SimSiam? j'assume que non, mais j'ai pas regardé le side-by-side du "SiameseArm" de pt-lightning --- t'as remarqué des trucs?

File: selfsupmotion/models/simsiam.py (hunk 1 above)
Comment (rjean) [reply selfsupmotion/models/simsiam.py:215]: Ça dépend du type de machine. On a environ 1 million de training samples. Pour un embedding size de 2048, on prend "juste" 8 gb de RAM. C'est pas un problème pour les machines avec lesquelles je travaille!

En ce moment, j'utilise des embeddings de 512, donc 2 gb de RAM. 

Est-ce que ça pourrait causer problème sur le cluster?

File: selfsupmotion/models/simsiam.py (hunk 2 above)
Comment (rjean) [reply selfsupmotion/models/simsiam.py:279]: Absolument, on va surement vouloir désactiver ça dans les "run" sur le cluster. Command line ou bien YAML hyper-paramètres?

**Comment by rjean** [comment]
Salut,

Pour la propreté du code, j'aurais tendance à prioriser les résultats. On pourrait faire un exercice de mise au propre plus tard vers la fin du projet pour bien "dérisquer" le projet. Si tu pense que ça peut nous rattraper avant la fin, alors mettons l'effort extra tout de suite. Au final, l'important pour moi soit que tu sois à l'aise avec la qualité du code. 

Pour ce qui est des différences de modèle, c'est principalement:
- Nombre de couche du projecteur. (3 plutôt que 2).
- Dimension des couches du projecteur et prédicteur. (Le projecteur n'a pas de bottleneck, mais le prédicteur en a un).
- Le fait d'utiliser des "bias" sur les couches linéaires.
- Le fait d'utiliser du "batch norm" à la sortie de la dernière couche linéaire du projecteur et prédicteur (si ma mémoire est bonne).

Le changement le plus important semble être l'asymétrie des "learning" rates. Ça change complétement les courbes d'apprentissage, je dois encore confirmer que ça a un impact important sur les représentations.

**Comment by plstcharles** [comment]
OK ça roule, je teste ça dès que possible, et si tout est beau, je merge après!

**Comment by plstcharles** [comment]
ya une modif dans simsiam.py qui fait crasher le checkpointing de pytorch-lightning quand on essaie d'entraîner directement par régression de kpts, ça semble venir d'un modèle... je vais réessayer de débugger demain pour débloquer le merge, à suivre!

**Comment by rjean** [comment]
Je n'avais pas testé pour le keypoints. Je viens de rendre la variable de "early stopping" configurable dans les hyper-paramètres. Aussi par défaut on va prendre "val_loss", mais dans mes configurations je vais prendre "val_accuracy". Ça devrait éviter le problème que tu as eu! Je teste l'entrainement par keypoints aujourd'hui sur ma machine, je devrais avoir compléter quelques "epochs" d'ici la fin de la journée.
